projects/power_distribution/src/pd_fan_ctrl.c

```diff
@@ -0,0 +1,177 @@
+    prv_fan_overtemp_callback(fan_overtemp_flags, v_dcdc, v_enc_vent);
+    *fan_speed_dcdc = 100;
+    *fan_speed_enc_vent = 100;
+    return;
+  }
+
+  // Approximate with linear temp/resistance plot of resistances
+  double dcdc_fraction =
+      (double)(dcdc_therm_res - FAN_OVERTEMP_RES) / (FAN_ENABLE_RES - FAN_OVERTEMP_RES);
+  double enc_vent_fraction =
+      (double)(enc_vent_therm_res - FAN_OVERTEMP_RES) / (FAN_ENABLE_RES - FAN_OVERTEMP_RES);
+  *fan_speed_dcdc = (uint8_t)((1.00 - dcdc_fraction) * 100);
+  *fan_speed_enc_vent = (uint8_t)((1.00 - enc_vent_fraction) * 100);
+}
+
+// Rear pd read temp values and adjust fan speeds on given interval
+static void prv_rear_pd_fan_ctrl_update_speed(SoftTimerId id, void *context) {
+  uint16_t ref_reading = 0;
+  uint16_t therm_reading_enc_vent = 0;
+  uint16_t therm_reading_dcdc = 0;
+  uint8_t fan_speed_dcdc = 0;
+  uint8_t fan_speed_enc_vent = 0;
+
+  // read adc values
+  adc_read_converted(ADC_CHANNEL_REF, &ref_reading);
+  adc_read_converted_pin(s_fan_storage.enc_vent_therm_pin, &therm_reading_enc_vent);
+  adc_read_converted_pin(s_fan_storage.dcdc_therm_pin, &therm_reading_dcdc);
+
+  // convert adc temp vals to fan percentages, call error cb if either is overtemp
+  prv_temp_to_fan_percent(ref_reading, therm_reading_enc_vent, therm_reading_dcdc, &fan_speed_dcdc,
+                          &fan_speed_enc_vent);
+  // set speed of fans based on above value
+  adt7476a_set_speed(s_fan_storage.i2c, fan_speed_enc_vent, s_fan_storage.fan_pwm1,
+                     ADT_I2C_WRITE_ADDRESS);
+  adt7476a_set_speed(s_fan_storage.i2c, fan_speed_dcdc, s_fan_storage.fan_pwm2,
+                     ADT_I2C_WRITE_ADDRESS);
+
+  soft_timer_start_millis(REAR_FAN_CONTROL_REFRESH_PERIOD_MILLIS,
+                          &prv_rear_pd_fan_ctrl_update_speed, NULL, NULL);
+}
+
+// Rear pd read temp values and adjust fan speeds on given interval
+static void prv_front_pd_fan_ctrl_update_speed(SoftTimerId id, void *context) {
+  uint16_t pot_reading = 0;
+  uint16_t ref_reading = 0;
+  uint8_t front_fan_speed = 0;
+  adc_read_converted(ADC_CHANNEL_REF, &ref_reading);
+  adc_read_converted_pin(s_fan_storage.pot_pin, &pot_reading);
+  double ratio = ref_reading / 100.0;
+  // Convert adc reading to fan speed percent
+  front_fan_speed = pot_reading / ratio;
+  adt7476a_set_speed(s_fan_storage.i2c, front_fan_speed, s_fan_storage.fan_pwm1,
+                     ADT_I2C_WRITE_ADDRESS);
+  adt7476a_set_speed(s_fan_storage.i2c, front_fan_speed, s_fan_storage.fan_pwm2,
+                     ADT_I2C_WRITE_ADDRESS);
+  soft_timer_start_millis(FRONT_FAN_CONTROL_REFRESH_PERIOD_MILLIS,
+                          &prv_front_pd_fan_ctrl_update_speed, NULL, NULL);
+}
+
+// Initialize fan control for given power distro
+StatusCode pd_fan_ctrl_init(FanCtrlSettings *settings, bool is_front_pd) {
+  if (settings == NULL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  s_is_front_pd = is_front_pd;
+  I2CSettings i2c_settings = {
```

Comment:
This is a bit of a weird one. ATD7476A driver calls I2C init, but i2c is already init'd in main.c, I can leave it as a reinit with the same settings as in main.c (may cause harder debugging) or I can go in and refactor the adt7476A driver to not call i2c init, which it probably shouldn't be doing anyway

Reply:
The ADT7476A driver was actually changed in the BMS main commit the other day to not init i2c anymore, if you rebase on master the change should be there.